File: cpp/src/Elections.cpp
#include "Elections.h"
#include <algorithm>
#include <numeric>
#include <sstream>
#include <iomanip>

void Elections::addCandidate(string candidate)
{
    officialCandidates.push_back(candidate);
    candidates.push_back(candidate);
    votesWithoutDistricts.push_back(0);
    votesWithDistricts["District 1"].push_back(0);
    votesWithDistricts["District 2"].push_back(0);
    votesWithDistricts["District 3"].push_back(0);
}

void Elections::voteFor(string elector, string candidate, string electorDistrict)
{
    if (!withDistrict)
    {
        if (count(candidates.begin(), candidates.end(), candidate) > 0)
        {
            int index = find(candidates.begin(), candidates.end(), candidate) - candidates.begin();
            votesWithoutDistricts[index] = votesWithoutDistricts[index] + 1;
        }
        else
        {
            candidates.push_back(candidate);
            votesWithoutDistricts.push_back(1);
        }
    }
    else
    {
        if (votesWithDistricts.contains(electorDistrict))
        {
            vector<int>& districtVotes = votesWithDistricts[electorDistrict];
            if (count(candidates.begin(), candidates.end(), candidate) > 0)
            {
                int index = find(candidates.begin(), candidates.end(), candidate) - candidates.begin();
                districtVotes[index] = districtVotes[index] + 1;
            }
            else
            {
                candidates.push_back(candidate);
                for (auto &[district, votes] : votesWithDistricts)
                {
                    votes.push_back(0);
                }
                districtVotes[candidates.size() - 1] = districtVotes[candidates.size() - 1] + 1;
            }
        }
    }
}

map<string, string> Elections::results()
{
    auto format = [](const auto &param)
    {
        // This function is only needed if std::format is not implemented in your standard library
        // or if the locale support doesn't allow usage of fr-FR
        ostringstream s;
        s << std::fixed << std::setprecision(2) << param << "%";
        string result = s.str();
        // HACK: This is a hack to simulate the fr-FR locale
        // We know that only numbers need to be formatted with "," instead of ".", so just replace them
        replace(result.begin(), result.end(), '.', ',' );
        return result;
    };

    map<string, string> results;
    int nbVotes = 0;
    int nullVotes = 0;
    int blankVotes = 0;
    int nbValidVotes = 0;

    if (!withDistrict)
    {
        nbVotes = accumulate(votesWithoutDistricts.begin(), votesWithoutDistricts.end(), 0);
        for (int i = 0; i < officialCandidates.size(); i++)
        {
            int index = find(candidates.begin(), candidates.end(), officialCandidates[i]) - candidates.begin();
            nbValidVotes += votesWithoutDistricts[index];
        }

        for (int i = 0; i < votesWithoutDistricts.size(); i++)
        {
            float candidatResult = ((float)votesWithoutDistricts[i] * 100) / nbValidVotes;
            string candidate = candidates[i];
            if (count(officialCandidates.begin(), officialCandidates.end(), candidate) > 0)
            {
                results[candidate] = format(candidatResult);
            }
            else
            {
                if (candidates[i].size() == 0)
                {
                    blankVotes += votesWithoutDistricts[i];
                }
                else
                {
                    nullVotes += votesWithoutDistricts[i];
                }
            }
        }
    }
    else
    {
        for (auto entry : votesWithDistricts)
        {
            vector<int> districtVotes = entry.second;
            int nbVotesPerDistrict = accumulate(districtVotes.begin(), districtVotes.end(), 0);
            nbVotes += nbVotesPerDistrict;
        }

        for (int i = 0; i < officialCandidates.size(); i++)
        {
            int index = find(candidates.begin(), candidates.end(), officialCandidates[i]) - candidates.begin();
            for (auto entry : votesWithDistricts)
            {
                vector<int> districtVotes = entry.second;
                nbValidVotes += districtVotes[index];
            }
        }

        map<string, int> officialCandidatesResult;
        for (int i = 0; i < officialCandidates.size(); i++)
        {
            officialCandidatesResult[candidates[i]] = 0;
        }
        for (auto entry : votesWithDistricts)
        {
            vector<float> districtResult;
            vector<int> districtVotes = entry.second;
            for (int i = 0; i < districtVotes.size(); i++)
            {
                float candidateResult = 0;
                if (nbValidVotes != 0)
                    candidateResult = ((float)districtVotes[i] * 100) / nbValidVotes;
                string candidate = candidates[i];
                if (count(officialCandidates.begin(), officialCandidates.end(), candidate) > 0)
                {
                    districtResult.push_back(candidateResult);
                }
                else
                {
                    if (candidates[i].size() == 0)
                    {
                        blankVotes += districtVotes[i];
                    }
                    else
                    {
                        nullVotes += districtVotes[i];
                    }
                }
            }
            int districtWinnerIndex = 0;
            for (int i = 1; i < districtResult.size(); i++)
            {
                if (districtResult[districtWinnerIndex] < districtResult[i])
                    districtWinnerIndex = i;
            }
            officialCandidatesResult[candidates[districtWinnerIndex]] = officialCandidatesResult[candidates[districtWinnerIndex]] + 1;
        }
        for (int i = 0; i < officialCandidatesResult.size(); i++)
        {
            float ratioCandidate = ((float)officialCandidatesResult[candidates[i]]) / officialCandidatesResult.size() * 100;
            results[candidates[i]] = format(ratioCandidate);
        }
    }

    float blankResult = ((float)blankVotes * 100) / nbVotes;
    results["Blank"] = format( blankResult);

    float nullResult = ((float)nullVotes * 100) / nbVotes;
    results["Null"] = format(nullResult);

    vector<vector<string>> values;
    transform(list.begin(), list.end(), back_inserter(values), [](const auto &val)
              { return val.second; });

    vector<int> sizes;
    transform(values.begin(), values.end(), back_inserter(sizes), [](const auto &v)
              { return v.size(); });
    int nbElectors = accumulate(sizes.begin(), sizes.end(), 0);
    float abstentionResult = 100 - ((float)nbVotes * 100 / nbElectors);
    results["Abstention"] = format(abstentionResult);

    return results;
}

File: csharp/elections/Elections/Elections.cs
using System.Globalization;

namespace Elections;


public class Elections
{
    List<String> candidates = new List<String>();
    List<String> officialCandidates = new List<String>();
    List<int> votesWithoutDistricts = new List<int>();
    Dictionary<String, List<int>> votesWithDistricts;
    private Dictionary<String, List<String>> list;
    private bool withDistrict;

    public Elections(Dictionary<String, List<String>> list, bool withDistrict)
    {
        this.list = list;
        this.withDistrict = withDistrict;

        votesWithDistricts = new Dictionary<string, List<int>>();
        votesWithDistricts["District 1"] = new List<int>();
        votesWithDistricts["District 2"] = new List<int>();
        votesWithDistricts["District 3"] = new List<int>();
    }

    public void addCandidate(String candidate)
    {
        officialCandidates.Add(candidate);
        candidates.Add(candidate);
        votesWithoutDistricts.Add(0);
        votesWithDistricts["District 1"].Add(0);
        votesWithDistricts["District 2"].Add(0);
        votesWithDistricts["District 3"].Add(0);
    }

    public void voteFor(String elector, String candidate, String electorDistrict)
    {
        if (!withDistrict)
        {
            if (candidates.Contains(candidate))
            {
                int index = candidates.IndexOf(candidate);
                votesWithoutDistricts[index] = votesWithoutDistricts[index] + 1;
            }
            else
            {
                candidates.Add(candidate);
                votesWithoutDistricts.Add(1);
            }
        }
        else
        {
            if (votesWithDistricts.ContainsKey(electorDistrict))
            {
                List<int> districtVotes = votesWithDistricts[electorDistrict];
                if (candidates.Contains(candidate))
                {
                    int index = candidates.IndexOf(candidate);
                    districtVotes[index] = districtVotes[index] + 1;
                }
                else
                {
                    candidates.Add(candidate);
                    foreach(var (district, votes) in votesWithDistricts){
                        votes.Add(0);
                    }
                    districtVotes[candidates.Count - 1] = districtVotes[candidates.Count - 1] + 1;
                }
            }
        }
    }

    public Dictionary<String, String> results()
    {
        Dictionary<String, String> results = new Dictionary<String, String>();
        int nbVotes = 0;
        int nullVotes = 0;
        int blankVotes = 0;
        int nbValidVotes = 0;

        if (!withDistrict)
        {
            nbVotes = votesWithoutDistricts.Sum();
            for (int i = 0; i < officialCandidates.Count; i++)
            {
                int index = candidates.IndexOf(officialCandidates[i]);
                nbValidVotes += votesWithoutDistricts[index];
            }

            for (int i = 0; i < votesWithoutDistricts.Count; i++)
            {
                float candidatResult = ((float)votesWithoutDistricts[i] * 100) / nbValidVotes;
                String candidate = candidates[i];
                if (officialCandidates.Contains(candidate))
                {
                    results[candidate] = String.Format(CultureInfo.GetCultureInfo("fr-FR"), "{0:F2}%", candidatResult);
                }
                else
                {
                    if (String.IsNullOrEmpty(candidates[i]))
                    {
                        blankVotes += votesWithoutDistricts[i];
                    }
                    else
                    {
                        nullVotes += votesWithoutDistricts[i];
                    }
                }
            }
        }
        else
        {
            foreach (var entry in votesWithDistricts)
            {
                List<int> districtVotes = entry.Value;
                nbVotes += districtVotes.Sum();
            }

            for (int i = 0; i < officialCandidates.Count; i++)
            {
                int index = candidates.IndexOf(officialCandidates[i]);
                foreach (var entry in votesWithDistricts)
                {
                    List<int> districtVotes = entry.Value;
                    nbValidVotes += districtVotes[index];
                }
            }

            Dictionary<String, int> officialCandidatesResult = new Dictionary<String, int>();
            for (int i = 0; i < officialCandidates.Count; i++)
            {
                officialCandidatesResult[candidates[i]] = 0;
            }
            foreach (var entry in votesWithDistricts)
            {
                List<float> districtResult = new List<float>();
                List<int> districtVotes = entry.Value;
                for (int i = 0; i < districtVotes.Count; i++)
                {
                    float candidateResult = 0;
                    if (nbValidVotes != 0)
                        candidateResult = ((float)districtVotes[i] * 100) / nbValidVotes;
                    String candidate = candidates[i];
                    if (officialCandidates.Contains(candidate))
                    {
                        districtResult.Add(candidateResult);
                    }
                    else
                    {
                        if (String.IsNullOrEmpty(candidates[i]))
                        {
                            blankVotes += districtVotes[i];
                        }
                        else
                        {
                            nullVotes += districtVotes[i];
                        }
                    }
                }
                int districtWinnerIndex = 0;
                for (int i = 1; i < districtResult.Count; i++)
                {
                    if (districtResult[districtWinnerIndex] < districtResult[i])
                        districtWinnerIndex = i;
                }
                officialCandidatesResult[candidates[districtWinnerIndex]] = officialCandidatesResult[candidates[districtWinnerIndex]] + 1;
            }
            for (int i = 0; i < officialCandidatesResult.Count; i++)
            {
                float ratioCandidate = ((float)officialCandidatesResult[candidates[i]]) / officialCandidatesResult.Count * 100;
                results[candidates[i]] = String.Format(CultureInfo.GetCultureInfo("fr-FR"), "{0:F2}%", ratioCandidate);
            }
        }

        float blankResult = ((float)blankVotes * 100) / nbVotes;
        results["Blank"] = String.Format(CultureInfo.GetCultureInfo("fr-FR"), "{0:F2}%", blankResult);

        float nullResult = ((float)nullVotes * 100) / nbVotes;
        results["Null"] = String.Format(CultureInfo.GetCultureInfo("fr-FR"), "{0:F2}%", nullResult);

        int nbElectors = list.Values.Select(v => v.Count).Sum();
        float abstentionResult = 100 - ((float)nbVotes * 100 / nbElectors);
        results["Abstention"] = String.Format(CultureInfo.GetCultureInfo("fr-FR"), "{0:F2}%", abstentionResult);

        return results;
    }
}
